server/core/include/scoped_permission.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef IRODS_SCOPED_PERMISSION_HPP
+#define IRODS_SCOPED_PERMISSION_HPP
+
+/// \file
+
+#include "rcConnect.h"
+#include "scoped_privileged_client.hpp"
+#include "rodsLog.h"
+
+#define IRODS_FILESYSTEM_ENABLE_SERVER_SIDE_API
+#include "filesystem.hpp"
+
+#include <utility>
+
+namespace irods::experimental
+{
+    /// This class provides a convenient RAII-style mechanism for setting permissions
+    /// on a collection or data object for the duration of a scoped block.
+    ///
+    /// Instances of this class are not copyable or moveable.
+    ///
+    /// \since 4.2.9
+    class scoped_permission
+    {
+    public:
+        /// Constructs a scoped_permission object and sets the requested permissions on the given
+        /// collection or data object to the client in RsComm::clientUser.
+        ///
+        /// \param[in] _comm The server communication object.
+        /// \param[in] _path A logical path representing a collection or data object.
+        /// \param[in] _perm The permissions to set on the collection or data object for the client.
+        ///
+        /// \exception irods::experimental::filesystem::filesystem_error
+        scoped_permission(RsComm& _comm, const filesystem::path& _path, filesystem::perms _perm)
+            : comm_{_comm}
+            , path_{_path}
+            , old_perm_{filesystem::perms::null}
+        {
+            namespace fs = filesystem;
+            
+            // Elevate privileges so that we can see the permissions.
+            scoped_privileged_client spc{comm_};
+
+            // Capture the client's current permissions if any.
+            const auto s = fs::server::status(comm_, path_);
+            for (auto&& entity : s.permissions()) {
+                if (entity.name == comm_.clientUser.userName &&
+                    entity.zone == comm_.clientUser.rodsZone &&
+                    entity.prms != fs::perms::inherit &&
+                    entity.prms != fs::perms::noinherit)
+                {
+                    old_perm_ = entity.prms;
+                    break;
+                }
+            }
+
+            // Set the client's requested permissions on the collection or data object.
+            fs::server::permissions(comm_, path_, comm_.clientUser.userName, _perm);
```

Comment:
this line will throw because the fsys library doesn't support the `ADMIN_KW`.

the fix for this is to either use the C API directly or provide a way to set the `ADMIN_KW` through the fsys library.